daemon/containerd/image_exporter.go

```diff
@@ -41,33 +94,133 @@ func (i *ImageService) ExportImage(ctx context.Context, names []string, outStrea
 		if err != nil {
-			// TODO(thaJeztah): remove this log or change to debug once we can; see https://github.com/moby/moby/pull/43822#discussion_r937502405
-			logrus.WithError(err).WithField("image", img.Name).Debug("failed to check if image is unpacked")
-			continue
+			logrus.WithError(err).WithField("image", img.Name).Debug("failed to get image platforms")
+			return errdefs.Unknown(err)
+		}
+
+		name := img.Name
+		if named, err := reference.ParseNormalizedNamed(img.Name); err == nil {
+			name = reference.FamiliarName(named)
 		}
 
-		if !unpacked {
-			err := platformImg.Unpack(ctx, i.snapshotter)
+		for _, platform := range allPlatforms {
+			logger := logrus.WithFields(logrus.Fields{
+				"platform": platform,
+				"image":    name,
+			})
+			platformImg := containerd.NewImageWithPlatform(i.client, img, cplatforms.OnlyStrict(platform))
+
+			unpacked, err := platformImg.IsUnpacked(ctx, i.snapshotter)
 			if err != nil {
-				// TODO(thaJeztah): remove this log or change to debug once we can; see https://github.com/moby/moby/pull/43822#discussion_r937502405
-				logrus.WithError(err).WithField("image", img.Name).Warn("failed to unpack image")
-				return errors.Wrap(err, "failed to unpack image")
+				logger.WithError(err).Debug("failed to check if image is unpacked")
+				continue
+			}
+
+			if !unpacked {
+				err = platformImg.Unpack(ctx, i.snapshotter)
+
+				if err != nil {
+					return errdefs.System(err)
+				}
 			}
+			logger.WithField("alreadyUnpacked", unpacked).WithError(err).Debug("unpack")
 		}
+
+		fmt.Fprintf(progress, "Loaded image: %s\n", name)
 	}
 	return nil
 }
+
+// getBestDescriptorForExport returns a descriptor which only references content available locally.
+// The returned descriptor can be:
+// - The same index descriptor - if all content is available
+// - Platform specific manifest - if only one manifest from the whole index is available
+// - Reduced index descriptor - if not all, but more than one manifest is available
+//
+// The reduced index descriptor is stored in the content store and may be garbage collected.
+// It's advised to pass a context with a lease that's long enough to cover usage of the blob.
+func (i *ImageService) getBestDescriptorForExport(ctx context.Context, indexDesc ocispec.Descriptor) (ocispec.Descriptor, error) {
+	none := ocispec.Descriptor{}
+
+	if !containerdimages.IsIndexType(indexDesc.MediaType) {
+		err := fmt.Errorf("index/manifest-list descriptor expected, got: %s", indexDesc.MediaType)
+		return none, errdefs.InvalidParameter(err)
+	}
+	store := i.client.ContentStore()
+	children, err := containerdimages.Children(ctx, store, indexDesc)
+	if err != nil {
+		if cerrdefs.IsNotFound(err) {
+			return none, errdefs.NotFound(err)
+		}
+		return none, errdefs.System(err)
+	}
+
+	// Check which platform manifests have all their blobs available.
+	hasMissingManifests := false
+	presentManifests := []ocispec.Descriptor{}
```

Comment:
Nit: since `presentManifests` is marshaled to JSON iff it is nonempty, [it should be initialized to `nil`.](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices)
```suggestion
	var presentManifests []ocispec.Descriptor
```

Reply:
Done, thanks!